[WPE] Use ASharedMemory when targeting Android

aperezdc opened the pull request:
#### 26eadedcaa761b153ddbbe473d932c240d95cbfe
<pre>
[WPE] Use ASharedMemory when targeting Android
<a href="https://bugs.webkit.org/show_bug.cgi?id=290210">https://bugs.webkit.org/show_bug.cgi?id=290210</a>

Reviewed by Carlos Garcia Campos.

When targeting Android use the ASharedMemory API to create shared memory
buffers and to adjust their protection bits. OS(ANDROID) guards ensure
that only the new implementation gets built.

* Source/WebCore/SourcesWPE.txt: List new SharedMemoryAndroid.cpp source.
* Source/WebCore/platform/android/SharedMemoryAndroid.cpp: Added, based
on the code from SharedMemoryUnix.cpp but using ASharedMemory instead.
(WebCore::SharedMemoryHandle::SharedMemoryHandle):
(WebCore::SharedMemoryHandle::releaseHandle):
(WebCore::accessModeMMap):
(WebCore::createSharedMemory):
(WebCore::SharedMemory::allocate):
(WebCore::SharedMemory::map):
(WebCore::SharedMemory::wrapMap):
(WebCore::SharedMemory::~SharedMemory):
(WebCore::SharedMemory::createHandle):
* Source/WebCore/platform/unix/SharedMemoryUnix.cpp:

Canonical link: <a href="https://commits.webkit.org/293579@main">https://commits.webkit.org/293579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ecdf72dd06b340372503845f525126e1116f8c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99336 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9236 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104467 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49937 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101377 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19274 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27419 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75600 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32706 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102343 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14651 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55960 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14445 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7666 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49297 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84370 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7753 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106824 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26450 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19291 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84560 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26812 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85887 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84072 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21326 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28743 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6429 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20190 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26390 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31590 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26210 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29523 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27777 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

csaavedra commented on Source/WebCore/platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -76,11 +82,11 @@ static inline int accessModeMMap(SharedMemory::Protection protection)
 }
 
-static UnixFileDescriptor createSharedMemory()
+static UnixFileDescriptor createSharedMemory([[maybe_unused]] size_t size)
```
Isn't `UNUSED_PARAM(variable)` the standard way of doing this?

aperezdc replied:
New code has been introducing `[[maybe_unused]]` which [exists since C++17](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused). There was a [discussion in webkit-dev](https://lists.webkit.org/pipermail/webkit-dev/2024-January/032609.html) and the consensus seemed to prefer the standard method over our custom macros.

csaavedra commented on Source/WebCore/platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -176,7 +195,7 @@ SharedMemory::~SharedMemory()
 
-auto SharedMemory::createHandle(Protection) -> std::optional<Handle>
+auto SharedMemory::createHandle([[maybe_unused]] Protection protection) -> std::optional<Handle>
 {
```
Same here.

csaavedra reviewed (commented):
Minor comments.

carlosgcampos commented on Source/WebCore/platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -28,6 +28,10 @@
 #include "config.h"
 #include "SharedMemory.h"
 
+#if defined(__ANDROID__) || defined(ANDROID)
```
We use this in other files already, could we add a `OS(ANDROID)` definition?

aperezdc replied:
I have made a separate patch to add `OS(ANDROID)` and change the existing preprocessor guards: https://github.com/WebKit/WebKit/pull/43460 &mdash; I'll change the guards in this and other upcoming PRs accordingly.

carlosgcampos commented on Source/WebCore/platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -28,6 +28,10 @@
 #include "SharedMemory.h"
 
+#if defined(__ANDROID__) || defined(ANDROID)
+#define HAVE_ASHAREDMEMORY 1
```
Why do we need this? It's enabled always for android, why don't we just check `OS(ANDROID)` everywhere?

aperezdc replied:
To write less down below. I'll change it to `OS(ANDROID)`.

carlosgcampos commented on Source/WebCore/platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -76,11 +82,11 @@ static inline int accessModeMMap(SharedMemory::Protection protection)
     int fileDescriptor = -1;
 
-#if HAVE(LINUX_MEMFD_H)
+#if HAVE(LINUX_MEMFD_H) && !HAVE(ASHAREDMEMORY)
```
This file is small enough and already has several ifdefs, could we just add SharedMemoryAndroid.cpp?

aperezdc replied:
The new file would be mostly a copy. I can give it a try and see how that looks like.

carlosgcampos commented on Source/WebCore/platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -108,29 +110,42 @@ static UnixFileDescriptor createSharedMemory()
 RefPtr<SharedMemory> SharedMemory::allocate(size_t size)
 {
-    auto fileDescriptor = createSharedMemory();
+    auto fileDescriptor = createSharedMemory(size);
```
So, we are adding ifdefs for all createSharedMemory code, why not doing something like:

```
#if OS(ANDROID)
static UnixFileDescriptor createSharedMemory(size_t size)
{
    int fileDescriptor;
    auto name = makeString("/WK2SharedMemory."_s, cryptographicallyRandomNumber<unsigned>()).utf8();
    do {
        fileDescriptor = ASharedMemory_create(name.data(), size);
    } while (fileDescriptor == -1 && errno == EINTR);

    return UnixFileDescriptor { fileDescriptor, UnixFileDescriptor::Adopt };
}
#else
static UnixFileDescriptor createSharedMemory()
{
    <current code>
}
#endif
```

carlosgcampos replied:
I'm assuming we don't need the 10 retries in android, but if we do, we could use the same loop here. This reduces the ifdefs and makes the code a lot easier to read without having to use a separate file for android.

carlosgcampos commented on Source/WebCore/platform/android/SharedMemoryAndroid.cpp:
```diff
@@ -0,0 +1,169 @@
+{
+    const auto name = makeString("/WK2SharedMemory."_s, cryptographicallyRandomNumber<unsigned>());
+    CString tempName = name.utf8();
+    int fileDescriptor = ASharedMemory_create(tempName.data(), size);
```
We don't need the `tempName` here, we use that in the unix implementation because we try 10 times in a loop. Here we can just do `name.utf8().data()`

aperezdc replied:
Right, I'll apply the suggestion before landing.

carlosgcampos commented on Source/WebCore/platform/unix/SharedMemoryUnix.cpp:
```diff
@@ -28,7 +28,7 @@
 #include "SharedMemory.h"
 
-#if USE(UNIX_DOMAIN_SOCKETS)
+#if !OS(ANDROID) && USE(UNIX_DOMAIN_SOCKETS)
```
It doesn't really matter but I think it's clearer `#if USE(UNIX_DOMAIN_SOCKETS) && !OS(ANDORID)`

aperezdc replied:
Sure, I'll switch the order of the checks before landing.

aperezdc committed:
[WPE] Use ASharedMemory when targeting Android
https://bugs.webkit.org/show_bug.cgi?id=290210

Reviewed by Carlos Garcia Campos.

When targeting Android use the ASharedMemory API to create shared memory
buffers and to adjust their protection bits. OS(ANDROID) guards ensure
that only the new implementation gets built.

* Source/WebCore/SourcesWPE.txt: List new SharedMemoryAndroid.cpp source.
* Source/WebCore/platform/android/SharedMemoryAndroid.cpp: Added, based
on the code from SharedMemoryUnix.cpp but using ASharedMemory instead.
(WebCore::SharedMemoryHandle::SharedMemoryHandle):
(WebCore::SharedMemoryHandle::releaseHandle):
(WebCore::accessModeMMap):
(WebCore::createSharedMemory):
(WebCore::SharedMemory::allocate):
(WebCore::SharedMemory::map):
(WebCore::SharedMemory::wrapMap):
(WebCore::SharedMemory::~SharedMemory):
(WebCore::SharedMemory::createHandle):
* Source/WebCore/platform/unix/SharedMemoryUnix.cpp:

Canonical link: https://commits.webkit.org/293579@main
